liquidhaskell-boot/src/Language/Haskell/Liquid/Bare/Axiom.hs

```diff
@@ -348,6 +356,26 @@ makeAssumeType allowTC tce lmap dm sym mbT v def
     ref        = F.Reft (F.vv_, F.PAtom F.Eq (F.EVar F.vv_) le)
     mkErr s    = ErrHMeas (sourcePosSrcSpan $ loc sym) (pprint $ val sym) (PJ.text s)
     bbs        = filter isBoolBind xs
+
+    -- rTypeSortExp produces monomorphic sorts from polymorphic types.
+    -- As an example, for 
+    -- id :: a -> a ... id x = x 
+    -- we got: 
+    -- define id (x : a#foobar) : a#foobar = { (x : a#foobar) }
+    -- Using FObj instead of a real type variable (FVar i) This code solves the
+    -- issue by creating a sort substitution that replaces those "fake" type variables
+    -- with actual ones.
+    -- define id (x : @-1) : a@-1 = { (x : a@-1) }
+    (tyVars, _) = Ghc.splitForAllTyCoVars τ
+    sortSub     = F.mkSortSubst $ zip (fmap F.symbol tyVars) (F.FVar <$> freeSort)
+    -- We need sorts that aren't polluted by rank-n types, we can't just look at
+    -- the term to determine statically what is the "maximum" sort bound ex:
+    -- freeSort = [1 + (maximum $ -1 : F.sortAbs out : fmap (F.sortAbs . snd) xts) ..] 
+    -- as some variable may be bound to something of rank-n type.  In
+    -- SortCheck.hs in fixpoint they just start at 42 for some reason.  I think
+    -- Negative Debruijn indices (levels :^)) are safer
+    freeSort    = [-1, -2 ..]
```

Comment:
These are the sorts produced for composition, according to `--save`:
```
define GHC.Base_LHAssumptions.comp
         ( lq1 : func(0 , [@(-1); @(-2)])
         , lq2 : func(0 , [@(-3); @(-1)])
         , lq3 : @(-3)
         ) : @(-2) = {((lq1 (lq2 lq3)))}
```
Are these correct? The meaning of negative sort variables is unclear to me.

What should be the sort of `comp`? Maybe

```
constant GHC.Base_LHAssumptions.comp
         : func (3,
             [ func(0 , [@(-1); @(-2)])
             ; func(0 , [@(-3); @(-1)])
             ; @(-3)
             ; @(-2)
             ])
```

Reply:
You need negative sort to prevent weird interactions with the rank2's, you need to renormalize them of you want to use func

Reply:
The normalized version should be `func(3, [func(0, [@0, @1]), func(0, [@2, @0]), @2, @1])`

Reply:
That's the same you should see for the constant in the fq file

Reply:
Alright, that makes sense to me.